service/history/chasm/context.go

```diff
@@ -0,0 +1,39 @@
+package chasm
+
+import (
+	"context"
+	"time"
+)
+
+type Context interface {
+	// Context is not binded to any component,
+	// so all methods needs to take in component as a parameter
+
+	// NOTE: component created in the current transaction won't have a ref
+	// this is a Ref to the component state at the start of the transition
+	Ref(Component) (ComponentRef, bool)
+	Now(Component) time.Time
+
+	// Intent() OperationIntent
+	// ComponentOptions(Component) []ComponentOption
+
+	getContext() context.Context
+}
+
+type MutableContext interface {
+	Context
+
+	AddTask(Component, TaskAttributes, any) error
+
+	// Add more methods here for other storage commands/primitives.
+	// e.g. HistoryEvent
+
+	// Get a Ref for the component
+	// This ref to the component state at the end of the transition
+	// Same as Ref(Component) method in Context,
+	// this only works for components that already exists at the start of the transition
+	//
+	// If we provide this method, then the method on the engine doesn't need to
+	// return a Ref
+	// NewRef(Component) (ComponentRef, bool)
```

Comment:
You could have a `CreationContext` that simply doesn't include this, and that and `MutableContext` both embed an un-exported base interface.